GLUSTERSR.py

```diff
@@ -0,0 +1,277 @@
+                os.symlink(self.linkpath, self.path)
+            except util.CommandException, inst:
+                if inst.code != errno.EEXIST:
+                    try:
+                        self.unmount(self.mountpoint, True)
+                    except GlusterFSException:
+                        util.logException('GlusterFSSR.unmount()')
+                    raise SR.SROSError(116,
+                                       "Failed to create GlusterFS SR. remote directory creation error: {}".format(
+                                           os.strerror(inst.code)))
+        self.detach(sr_uuid)
+
+    def delete(self, sr_uuid):
+        # try to remove/delete non VDI contents first
+        super(GlusterFSSR, self).delete(sr_uuid)
+        try:
+            if self.checkmount():
+                self.detach(sr_uuid)
+            self.mount()
+            if util.ioretry(lambda: util.pathexists(self.linkpath)):
+                util.ioretry(lambda: os.rmdir(self.linkpath))
+            self.unmount(self.mountpoint, True)
+        except util.CommandException, inst:
+            self.detach(sr_uuid)
+            if inst.code != errno.ENOENT:
+                raise SR.SROSError(114, "Failed to remove GlusterFS mount point")
+
+    # noinspection PyPep8Naming
+    def vdi(self, uuid, loadLocked=False):
+        if not loadLocked:
+            return GlusterFSFileVDI(self, uuid)
+        return GlusterFSFileVDI(self, uuid)
+
+
+class GlusterFSFileVDI(FileSR.FileVDI):
+    def attach(self, sr_uuid, vdi_uuid):
+        if not hasattr(self, 'xenstore_data'):
+            self.xenstore_data = {}
+        self.xenstore_data["storage-type"] = "glusterfs"
+        return super(GlusterFSFileVDI, self).attach(sr_uuid, vdi_uuid)
+
+    def generate_config(self, sr_uuid, vdi_uuid):
+        util.SMlog("SMBFileVDI.generate_config")
+        if not util.pathexists(self.path):
+            raise xs_errors.XenError('VDIUnavailable')
+        resp = {'device_config': self.sr.dconf,
+                'sr_uuid': sr_uuid,
+                'vdi_uuid': vdi_uuid,
+                'sr_sm_config': self.sr.sm_config,
+                'command': 'vdi_attach_from_config'}
+        # Return the 'config' encoded within a normal XMLRPC response so that
+        # we can use the regular response/error parsing code.
+        config = xmlrpclib.dumps(tuple([resp]), "vdi_attach_from_config")
+        return xmlrpclib.dumps((config,), "", True)
+
+    def attach_from_config(self, sr_uuid, vdi_uuid):
+        try:
+            if not util.pathexists(self.sr.path):
+                self.sr.attach(sr_uuid)
+        except:
+            util.logException("SMBFileVDI.attach_from_config")
+            raise xs_errors.XenError('SRUnavailable',
+                                     opterr='Unable to attach from config')
+
+
+if __name__ == '__main__':
+    SRCommand.run(GlusterFSSR, DRIVER_INFO)
+else:
+    SR.registerSR(GlusterFSSR)
```

Comment:
It's really a detail, but this file doesn't have an end of file newline, and the former does.